restart-script/restart-cluster

```diff
@@ -0,0 +1,235 @@
+  local actual_node_state
+
+  while [ "$expected_node_state" != "$actual_node_state" ]; do
+          actual_node_state=$(sshGenesis "curl -s http://$node_to_check:$port_to_check/cluster/info | jq '.[] | select( .ip == \"$node_to_check\" )' | jq '.state' -r")
+    echo -e "${pnk} Current node ($node_to_check) state: $actual_node_state. Expected state: $expected_node_state ${clr}"
+    sleep 1
+  done
+}
+
+stopServices() {
+  echo -e "${blu}Stopping node snd dag services${clr}"
+
+  while read -r ip; do
+    echo "Stopping services for: $ip"
+    ssh admin@$ip -o StrictHostKeyChecking=no -n 'sudo systemctl stop dag && sudo systemctl stop node'
+  done <$HOSTS_FILE
+}
+
+downloadRollbackSnapshot() {
+  echo -e "${blu}Downloading latest snapshot${clr}"
+
+  sshGenesis 'curl -s https://olubc3p4rc.execute-api.us-west-1.amazonaws.com/global-snapshots/latest | jq '.data.hash' -r > /home/admin/tessellation/rollback-snapshot'
+}
+
+prepareRollbackFlow(){
+  echo -e "${blu}Setting rollback flow for cluster${clr}"
+
+  latest_snapshot=`sshGenesis 'head -1 /home/admin/tessellation/rollback-snapshot'`
+  path_to_latest_snapshot="/home/admin/tessellation/data/snapshot/$latest_snapshot"
+
+  if sshGenesis "test -f $path_to_latest_snapshot" ; then
+    sshGenesis 'rm -f /home/admin/tessellation/start && ln -s /home/admin/tessellation/rollback /home/admin/tessellation/start'
+  else
+    echo -e "${red}Cannot find file with latest snapshot: ${grn}$latest_snapshot${clr}"
+    exit 1
+  fi
+}
+
+prepareGenesisFlow(){
+  echo -e "${blu}Setting genesis flow for cluster${clr}"
+
+  path_to_genesis="/home/admin/tessellation/genesis.csv"
+
+  if sshGenesis "[ -f $path_to_genesis ]" ; then
+    sshGenesis 'rm -f /home/admin/tessellation/start && ln -s /home/admin/tessellation/genesis /home/admin/tessellation/start'
+  else
+    echo -e "${red}Cannot find ${grn}genesis.csv${clr}"
+    exit 1
+  fi
+}
+
+prepareInitialValidator(){
+  echo -e "${blu}Setting initial validator${clr}"
+
+  sshGenesis 'rm -f /home/admin/tessellation/dag/start && ln -s /home/admin/tessellation/dag/initial-validator /home/admin/tessellation/dag/start'
+}
+
+restartNodeServices() {
+  echo -e "${blu}Restarting node services${clr}"
+
+  while read -r ip; do
+    echo "Restarting node service for: $ip"
+    ssh admin@$ip -o StrictHostKeyChecking=no -n 'sudo /home/admin/tessellation/restart'
+  done <$HOSTS_FILE
+}
+
+restartDagServices(){
+  echo -e "${blu}Restarting dag services${clr}"
+
+  while read -r ip; do
+    echo "Restarting dag service for: $ip"
+    ssh admin@$ip -o StrictHostKeyChecking=no -n 'sudo /home/admin/tessellation/dag/restart'
+  done <$HOSTS_FILE
+}
+
+joinGenesis() {
+  echo -e "${blu}Joining L0 nodes${clr}"
+  genesis_ip=`head -1 $HOSTS_FILE`
+
+  checkNodeState $genesis_ip 9000 "Ready"
+
+  while read ip; do
+    if [ "$ip" == "$genesis_ip" ]; then
+      echo "Skipping genesis node"
+    else
+      echo "Joining node: $ip to $genesis_ip"
+      sleep 15
```

Comment:
Not sure if you need to wait at all after the recent fixes? 🤔

Reply:
Ok, I think I will move this sleep before `while` loop like in function below, so other nodes have some more time to reach just to reach `ReadyToJoin` state. Unfortunately I cannot check such state like above for genesis node cause `node/state` endpoint is not public.